### PR TITLE
Partial Impale Support

### PR DESCRIPTION
**Pull request by baranio**
- Add local impale chance mods in weapons
- Add impale mods parsing support (chance, max stacks, effect, per
impale)
- Add 'while using' phrase support (equivalent to 'while affected by')
- Add basic impale calcs - chance, effectivenes, damage multiplier based
on (configurable) stacks
- Add impale options to configuration tab (stacks and have impaled
recently)
- Add impale skill stats to mods mappings
- Add prismatic jewel pride mods
- Add flagging 'affected by aura' state for Debuff Aura skills

### CODE REVIEW

**Comment by philippchistyakov** [comment]
Nice! Merge this! @Openarl

**Comment by besplash** [comment]
You also have to edit Data\3_0\ModCache.lua to add support for skill tree nodes

**Comment by baranio** [comment]
I'd imagine that cache should be rebuilt and commited once per release, but if @Openarl does express such a concern I should fix it.

File: Modules/CalcOffence-3_0.lua [hunk 1]
```diff
@@ -2002,6 +2008,42 @@ function calcs.offence(env, actor, activeSkill)
 				t_insert(breakdown.EnemyStunDuration, s_format("= %.2fs", output.EnemyStunDuration))
 			end
 		end
+
+        -- Calculate impale chance and modifiers
+		if canDeal.Physical and output.ImpaleChance > 0 then
+            skillFlags.impale = true
+            local impaleChance = output.ImpaleChance/100
```
Comment (chollinger) [inline Modules/CalcOffence-3_0.lua]:
```suggestion
            local impaleChance = m_min(output.ImpaleChance/100, 1)
```
Dunno if it's necessary with `combineStat` below, but you may wanna cap at 100% chance.

**Comment by chollinger** [comment]
This sets up everything except the actual damage calculation, yeah?

**Comment by baranio** [comment]
The chance should be capped at 100% when it is actually calculated. The
part you have pointed out merely changes the chance to decimal. But I agree
that double checking the cap wouldn't hurt.

Yeah I think it sets up pretty much everything , but I don't know whether
in a correct way (i.e. pride's buff wasn't taken into account, but I don't
know if my solution doesn't cause some issues)

DPS calculation is a bit tricky in my opinion, especially for dual wielding
as imaples from MH and OH would not occur on the same basis and more often
than not the monsters die even before imapale count reaches max count (and
thus imaples stack count option in configuration tab)

pt., 14 cze 2019, 19:09 użytkownik Cory Hollinger <notifications@github.com>
napisał:

**Comment by fleisc** [comment]
Impale chance crafted modifiers on 1H weapons are being applied to Impale chance for both weapons when dual wielding.

**Comment by baranio** [comment]
> Impale chance crafted modifiers on 1H weapons are being applied to Impale chance for both weapons when dual wielding.

Thanks for pointing that out

**Comment by LocalIdentity** [comment]
I figured out how to add support for the points on the tree to effect impale but am not sure how to get it working with the buffs from dread banner and the final DPS calculation, know how to get those working?

**Comment by fleisc** [comment]
Dread Banner's buffs needs to be marked as GlobalEffect in statmap in act_str.lua.  Should be able to mirror how this is done for other aura skills like Anger.

**Comment by LocalIdentity** [comment]
Got it working with Dread Banner now. Just need to fix the Master of Metal node on Champion to add the flat phys for each impale, have it currently working using shocked enemies killed recently instead of each Impale on the enemy (not sure how to get it working). Tried to push the changes but got an access denied message, so I'm not sure how to list them for others to see.

**Comment by Lumenblood** [comment]
The damage shouldn't be that hard to implement now, right? We already have the basic impale stats, now all it needs is scaling with the base physical damage at max stacks (or the given number of stacks) and finally added with total damage. The formula is something like this: 

100*Percent of physical dmg  * Impale chance * (1 + impale effect) * 0.1 * 5 ( + 2 with champion's Master of Metal)%  added with total damage.                                               {credits to u/Wanderlaars on Reddit}

I'd try to add it myself, but I know nothing about programming.

**Comment by monkey316** [comment]
Will this be merged at some point?

**Comment by jamison654321** [comment]
How can I add this to my POB?

**Comment by besplash** [comment]
> 
> 
> How can I add this to my POB?

Just use LocalIdentitys PoB fork

**Pull request closed without merging**